### ISSUE TITLE
How to Integrate Gauge Project to TeamCity

### ISSUE THREAD

**Issue by itsmemittal** [getgauge/gauge-csharp#45]

**Comment by sriv** [comment]
The documentation [here](http://getgauge.io/documentation/user/current/integration_with_ci/go.html) has steps to setup Gauge with GoCD(https://www.go.cd/). You can use this as a reference and set it up with teamcity as well.

**Comment by itsmemittal** [comment]
I have tried it out from documentation but on running build to CI it saying 'Failed to connect start gauge API' but if i am running same project with command line then working fine.

**Comment by itsmemittal** [comment]
![issue](https://cloud.githubusercontent.com/assets/3033471/11686023/9abeef10-9ea3-11e5-9445-7b1ccbbac442.png)

**Comment by sriv** [comment]
This could be because your build (of the target project) is failing on the build agent.

There are a few things to verify: 

- Your project isn't setup to restore nuget packages on build, which means that the reference to Gauge.CSharp.Lib is missing (looking at your repo, it seems like you have checked in the nupkg file alone). 
- Verify the eventlog on the build agent, this should have some details on what went wrong.

**Comment by itsmemittal** [comment]
I have set the project for Enable Restore Package on Build and committed nuget.exe to repository but getting same (above image ) message again

**Comment by itsmemittal** [comment]
![cmdoutput](https://cloud.githubusercontent.com/assets/3033471/11708389/d96c0068-9f34-11e5-8f2a-daa620e33f90.png)

When run in command prompt then saying failed to open gauge.properties and also executing specification

**Comment by sriv** [comment]
Those `Warnings` should not affect the execution, you should be able to see the reports. But the location does seem incorrect. 

The `gauge.properties` file is located at `C:\Program Files\Gauge\share\gauge`, but in your case, Gauge seems to try to read it from `C:\Program Files\Gauge\bin\share\gauge`, which is not right. 

The reason these are warnings, is because Gauge does have defaults that it falls back to, if it cannot read the values.

I have opened a separate issue to track the incorrect location : https://github.com/getgauge/gauge/issues/269

**Comment by itsmemittal** [comment]
But with cmd its executing when run through teamcity it failed with "failed
to connect start gauge api"

**Comment by sriv** [comment]
Oh sorry, I thought your screenshot was from the teamcity agent. 

Gauge plugins are installed per user, so in your CI setup, can you check if the user that you use to install gauge-csharp plugin and the user that teamcity uses to run the specs are same?

**Comment by itsmemittal** [comment]
Thanks Shrikant,

I am now able to run gauge commands with teamcity would you please favor me on one more question that how to get rid of web driver events from console output.


Like following how to get rid

[16:10:38][Step 1/1] 16:10:38.412 [INFO] 1449744038413	addons.xpi	DEBUG	getModTime: Recursive scan of {972ce4c6-7e08-4474-a285-3208198ce6fd}

[16:10:38][Step 1/1] 16:10:38.413 [INFO] 1449744038414	addons.xpi	DEBUG	New add-on fiddlerhook@fiddler2.com in winreg-app-global

[16:10:38][Step 1/1] 16:10:38.414 [INFO] 1449744038414	addons.xpi	DEBUG	getModTime: Recursive scan of fiddlerhook@fiddler2.com

**Comment by sriv** [comment]
Glad to know this is working well.

As for the webdriver events, I am afraid there isn't much that Gauge can do. Typically I've seen firefox with webdriver adds a good amount of debug logs when loading the plugins. One thing to try is to create a separate Firefox profile, with no addons created. 

I am going to close this issue, please open a new issue or use the googlegroups for anything else you need.

**Comment by Buthrakaur** [comment]
Hello @sriv I'm getting simmilar error on TeamCity (9.1.6), but I guess it's caused by some security issue. Do you know what could be wrong in my case? I'm installing the plugins in the same script by running `gauge --install-all` just before `gauge --env X --parallel .\specs`.

```
[16:14:57][Step 3/5] Compatible version of plugin csharp not found. Installing plugin csharp...
[16:14:57][Step 3/5] Gathering metadata for csharp
[16:14:59][Step 3/5] 
[16:14:59][Step 3/5] Downloading gauge-csharp-0.7.2.zip
[16:15:04][Step 3/5] .......................
[16:15:05][Step 3/5] Installing plugin csharp 0.7.2
[16:15:05][Step 3/5] Successfully installed plugin 'csharp'.
[16:15:05][Step 3/5] Compatible version of plugin html-report not found. Installing plugin html-report...
[16:15:05][Step 3/5] Gathering metadata for html-report
[16:15:06][Step 3/5] 
[16:15:06][Step 3/5] Downloading html-report-2.1.0-windows.x86_64.zip
[16:15:08][Step 3/5] ........................
[16:15:09][Step 3/5] Installing plugin html-report 2.1.0
[16:15:09][Step 3/5] Successfully installed plugin 'html-report'.
[16:15:39][Step 3/5] Failed to start gauge API: Timed out connecting to [::]:50541
```

**Comment by sriv** [comment]
@Buthrakaur - That is strange indeed. Can you please run a `gauge -v` just after `gauge --install-all` ? Ideally this should list C# as a gauge plugin.

Also, could you check the content of `gauge.log` usually located at `%APPDATA%\gauge\logs`?

**Comment by Buthrakaur** [comment]
@sriv : gauge -v on TC prints:
```
[07:58:27][Step 3/5] Gauge version: 0.4.0
[07:58:27][Step 3/5] 
[07:58:27][Step 3/5] Plugins
[07:58:27][Step 3/5] -------
[07:58:27][Step 3/5] csharp (0.7.2)
[07:58:27][Step 3/5] html-report (2.1.0)
[07:58:27][Step 3/5] xml-report (0.1.2)
```

I frist thought it's related to some security issue with opening port on the TC agent, so I added test of opening ports to beginning of my powershell gauge-start script:
```
@(5950, 666, 81, 25, 21) | 
    %{
        $Listener = [System.Net.Sockets.TcpListener]$_;
        $Listener.Start();
        Write-Host 'listening on port ' $_
        $Listener.Stop();
    }
```
The `TcpListener` starts listening on those ports without any problems, so the issue probably doesn't relate to problem with ports at all... I also tried to use `--api-port` and `--daemonize`, but the error is still the same.

Just to provide complete information - I don't have gauge installed on the TC agent. I use it in portable way - I have gauge `bin` and `share` folders commited in the git repository and set `GAUGE_ROOT` environment variable in my gauge launching script.

`gauge.log` file:
```
08:08:49.208 Compatible version of plugin csharp not found. Installing plugin csharp...
08:08:49.209 Gathering metadata for csharp
08:09:01.738 Downloading gauge-csharp-0.7.2.zip
08:09:29.359 Installing plugin csharp 0.7.2
08:09:29.447 Successfully installed plugin 'csharp'.
08:09:29.447 Compatible version of plugin html-report not found. Installing plugin html-report...
08:09:29.447 Gathering metadata for html-report
08:09:41.570 Downloading html-report-2.1.0-windows.x86_64.zip
08:09:57.626 Installing plugin html-report 2.1.0
08:09:57.687 Successfully installed plugin 'html-report'.
08:09:57.688 Compatible version of plugin xml-report not found. Installing plugin xml-report...
08:09:57.688 Gathering metadata for xml-report
08:09:58.715 Downloading xml-report-0.1.2-windows.x86_64.zip
08:10:00.920 Installing plugin xml-report 0.1.2
08:10:00.930 Successfully installed plugin 'xml-report'.
08:10:02.066 Checking updates...
08:11:01.015 Runner connection error: Timed out connecting to [::]:50362
08:11:01.015 Error while killing runner: invalid argument
08:11:01.015 Failed to start gauge API: Timed out connecting to [::]:50362
```

`api.log` file wasn't created at all.

**Comment by Buthrakaur** [comment]
@sriv more details on the issue - the TC agents run as a Local System account. When I try to run the same script on my machine under Local System (using `psexec -i -s Powershell.exe`) I get error dialog pointing to https://support.microsoft.com/en-us/kb/2715633

**Comment by sriv** [comment]
@Buthrakaur - can you please check what is the .NET framework version on the TC agent? Gauge.CSharp requires .NET 4.5.

**Comment by Buthrakaur** [comment]
@sriv: the .NET installation is ok, but the Local System account is causing the problem. Gauge runs without any problems when I run it under Administrator account directly on the TC agent (I'm connected to the agent over RDP).

**Comment by Buthrakaur** [comment]
And I'm able to simulate on my local machine too - when I switch to Local System account using `psexec -i -s Powershell.exe` and try to launch gauge tests, which just work under my user account, I get the error dialog and gauge fails with `Failed to start gauge API: Timed out connecting to [::]:13453`.

**Comment by sriv** [comment]
Thanks for isolating the issue @Buthrakaur , this is helpful. I am reopening this issue.

Could you also please look at your Windows Event Logs and see if there are any errors with stacktrace? If you can remove sensitive information and share the stacktrace, that would be great.

**Comment by Buthrakaur** [comment]
I'm not able to find anything in event log.

**Comment by Buthrakaur** [comment]
I'm able to run `Gauge.Csharp.Eunner.exe` from the commandline under Local System without the error dialog, so it looks like the problem must be caused not by the C# runner itself, but by the method how `Gauge.exe` starts the `Gauge.Csharp.Eunner.exe` process. How does actually this work?

**Comment by Buthrakaur** [comment]
@sriv : is it possible to start `gauge.exe` on some port without the csharp plugin and run `Gauge.Csharp.Runner.exe` manually? I could try it as a workaround...

**Comment by sriv** [comment]
@Buthrakaur - I can see this issue on my machine as well when I use `psexec -i -s powershell.exe` as you suggested. I am not sure why this is happening though :(.

On running `gauge.exe` and `Gauge.Csharp.Runner.exe` manually, I don't think that is possible, since the TCP port used to communicate is set by `gauge.exe` and passed on to the runner.

**Comment by Buthrakaur** [comment]
@sriv isn't is possible to enforce the `gauge.exe` port using `GAUGE_API_PORT` and `GAUGE_INTERNAL_PORT` variables (http://getgauge.io/documentation/technical/current/types_of_plugins/communication.html)? I tried it, but the problem was `gauge.exe` was rejecting to start when I had no runner registered in the manifest, so I couldn't try it.

**Comment by sriv** [comment]
@Buthrakaur - the interaction between `gauge.exe

**Comment by Buthrakaur** [comment]
@sriv I see... So workaround I had on my mind isn't possible :( The problem is probably in the GO implementation (it probably starts the .NET process in some special way), but I don't know GO at all to be able to diagnose or even fix the issue. The thing is TeamCity Windows agent default settings is to run under Local System (and it kinda makes sense), so gauge integration with TC is blocked by this issue. I'll highly appreciate any help with this, because this is show-stopper for our team..

**Comment by sriv** [comment]
@Buthrakaur - I understand the criticality of this one. I am working on this as we discuss. Appreciate your  effort and patience on this one.

-------- [Tech Notes] ---------------

Some internals (just FYI, since you've dug it up so far)

Gauge doesn't contain the launch mechanism of the runner, the runner's metadata does. In this case if you notice the `%APPDATA%\gauge\plugins\csharp\0.7.2\csharp.json` it should be like [this file](https://github.com/getgauge/gauge-csharp/blob/master/Runner/csharp.json).

When Gauge requires the runner, it invokes the [command specified](https://github.com/getgauge/gauge-csharp/blob/master/Runner/csharp.json#L11) in the metadata.

**Comment by Buthrakaur** [comment]
@sriv yes - I found  `csharp.json` :) But the `gauge.exe` code actually starts the process (https://github.com/getgauge/gauge/blob/7df4132b31e37a2b318bb58603501ec9b0e949e2/runner/runner.go#L102) and it looks like the process is being started in some unusual way, which is the root cause of the issue... or at least that's my guess :)

Thanks for your kind support.

**Comment by mahendrakariya** [comment]
Like @sriv said, it just starts the command mentioned in `csharp.json` and sets the Gauge project as the working directory for the command. So it doesn't look like that's an issue. 

The code for `common.ExecuteCommand()` is [over here](https://github.com/getgauge/common/blob/master/common.go#L484), just in case you would like to dig deeper.

**Comment by sriv** [comment]
@Buthrakaur - I think I have a workaround.

In your teamcity agent, can you do the following?

- Install your plugins to a [custom location](http://getgauge.io/documentation/user/current/plugins/installation.html#custom-plugin-install-location). i.e. set `GAUGE_HOME` to some location with liberal privileges for all users and run `gauge --install-all`.

- Run `gauge specs` as your normal Local System account.

I believe under Local System, the `%APPDATA%` points to `C:\Windows\system32\config\systemprofile\AppData` for Windows 7, WIndows Server 2008 and above. `gauge --install csharp` copies the language runner to `%APPDATA%\gauge\plugins` and this is causing some security issues.

I am still unclear on the exact cause and I don't know what is the permanent fix yet, will continue to investigate.

In the meantime, can you please check if the `GAUGE_HOME` workaround works?

**Comment by Buthrakaur** [comment]
@sriv: it looks promising, but I can't make gauge find spec files now. Really strange issue - could it relate to `GAUGE_HOME` variable?
```
[09:02:41][Step 4/6] $env:GAUGE_HOME C:\opt\app\BuildAgent\work\5ee4a55bac2c4ce0\src\Acc.Tests\bin
...
[09:02:54][Step 4/6] Gauge.CSharp.Runner.SandboxFactory Creating a Sandbox in: C:\opt\app\BuildAgent\work\5ee4a55bac2c4ce0\src\Acc.Tests\bin\Release
[09:02:58][Step 4/6] No specifications found in C:\opt\app\BuildAgent\work\5ee4a55bac2c4ce0\src\Acc.Tests\specs.
```
`ls C:\opt\app\BuildAgent\work\5ee4a55bac2c4ce0\src\Acc.Tests\specs -recurse` shows:
```
[09:02:10][Step 3/6]     Directory: C:\opt\app\BuildAgent\work\5ee4a55bac2c4ce0\src\Acc.Tests\specs\x\y\z
[09:02:10][Step 3/6] 
[09:02:10][Step 3/6] 
[09:02:10][Step 3/6] Mode                LastWriteTime     Length Name                              
[09:02:10][Step 3/6] ----                -------------     ------ ----                              
[09:02:10][Step 3/6] -a---          3.5.2016      9:01       1642 spec1.md  
```

**Comment by Buthrakaur** [comment]
@sriv my mistake - I had a typo in tags filter. Everything is working now - thanks a lot for your kind support :)
![image](https://cloud.githubusercontent.com/assets/126757/14977698/9381a22a-1116-11e6-9694-051fb65af813.png)

**Comment by sriv** [comment]
@Buthrakaur - neat, thanks for confirming. I am closing this issue, but I will be looking out for solutions that do not require `GAUGE_HOME` to be set.

**Comment by Buthrakaur** [comment]
@sriv : you could also add some information about this problem to http://getgauge.io/documentation/user/current/integration_with_ci/index.html

**Comment by sriv** [comment]
@Buthrakaur - yes, was thinking about it, perhaps since this is an issue reported on the use of Gauge with TeamCity, a section on TeamCity would make sense. What do you say?

**Comment by bkim54** [comment]
@sriv @Buthrakaur is there any official documentation on setting gauge up with TC? or if you could give me some pointers.

Thanks!

**Comment by Buthrakaur** [comment]
Hi @bkim54 ,
we're using portable version of Gauge, set few environment variables and run the tests using this Powershell script: https://gist.github.com/Buthrakaur/eb3b0022223d993778fe7cd6141379c4 . Than we have few things setup on teamcity:
- add Gauge HTML report as a build artifact
- added project build report tab with Gauge HTML report artifact
- import Gauge XML report using Build Feature JUnit XML Report Processing

**Comment by bkim54** [comment]
hey @Buthrakaur thanks, what do you mean by a portable version of gauge?

**Comment by Buthrakaur** [comment]
https://github.com/getgauge/gauge/releases/ => *.zip instead of exe
installer.

**Comment by bkim54** [comment]
@Buthrakaur  thanks i'll give it a shot
